server/etcdserver/v3_server.go

```diff
@@ -768,16 +769,22 @@ func (s *EtcdServer) requestCurrentIndex(leaderChangedNotifier <-chan struct{},
 		return 0, err
 	}
 
-	confirmedIndex, err := s.readReadIndexResponse(leaderChangedNotifier, requestId)
-	if err != nil {
-		return 0, err
-	}
-	return confirmedIndex, nil
-}
-
-func (s *EtcdServer) readReadIndexResponse(leaderChangedNotifier <-chan struct{}, requestId uint64) (uint64, error) {
 	lg := s.Logger()
+	errorTimer := time.NewTimer(s.Cfg.ReqTimeout())
```

Comment:
I don't understand it... 

Would following approach work: 
  - retryTimer instead of errorTimer. 
  - both selects merged into a single select.
?

Reply:
First, I've decided to use 2 selects instead of one because of [this article](https://dev.to/julianchu/go-for-select-with-timer-1ah8#timer). In short: it is to guard against picking `case` at random when multiple channels are unblocked. It seemed reasonable when writing code, however now I can't say any good reason to have it this way, so I'll move it to single select.

Second think is "retryTimer instead of errorTimer". I'm not sure if I understand you correctly, but if you suggest having retryTimer (one that measure 500ms) outside of for loop and putting errorTimer (one that measure 7s) in `<-time.After(...)`, then it would not work as we need to have single errorTimer across potentially multiple retryTimer.

Reply:
A.d. 1.  I see. But I think we don't really care which branch will be taken if it happens that multiple activate exactly at the same time. So would go for merging. 

A.d. 2: You are right. Potentially you can use top-level 'ticker' (time.NewTicker) to get periodic notification to refresh the request. The benefit is that it automatically will cancel goroutine.

Reply:
A.d. 2:
I see `Timer` more intuitive than `Ticker` here, as "ping me after 500ms" describes concept of timeout better than "ping me every 500ms". However I don't understand below part
> The benefit is that it automatically will cancel goroutine

Could you elaborate?

Reply:
<-time.After(readIndexRetryTime) 

under the cover starts go-routine the sleep's some time and populates the channel. 

If the select exits for another reason, the goroutne still exists for 'up to 500ms' to populate the channel (that no-one is waiting for).

Reply:
Good point. I've changed it to single timer initialised outside of loop and I use `retryTimer.Reset(readIndexRetryTime)` to refresh it.